Fix: Fixed an issue where it didn't work to map certain keys to Actions

XTorLukas opened the pull request:
<!-- 
🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨
I ACKNOWLEDGE THE FOLLOWING BEFORE PROCEEDING:
1. PR may be deleted if it is not following the template
2. Try not to make duplicates. Do a quick search before posting
3. Add a clear title starting with "Feature:" or "Fix:"
-->

**Resolved / Related Issues**

To prevent extra work, all changes to the Files codebase must link to an approved issue marked as `Ready to build`. Please insert the issue number following the hashtag with the issue number that this Pull Request resolves.
- #15403

**Steps used to test these changes**

Stability is a top priority for Files and all changes are required to go through testing before being merged into the repo. Please include a list of steps that you used to test this PR.

1. Testing the correct mapping of all problem keys

0x5bfa commented on src/Files.App/Data/Commands/HotKey/HotKey.cs:
```diff
@@ -332,9 +332,21 @@ private static string GetLocalizedKey(string key)
 			var state = new byte[256];
-			_ = Win32PInvoke.ToUnicode((uint)key, 0, state, buffer, 256, 0);
+
+			if (!Win32PInvoke.GetKeyboardState(state))
+			{
+				return buffer.ToString();
+			}
```
```suggestion
			if (!Win32PInvoke.GetKeyboardState(state))
				return buffer.ToString();
```

0x5bfa commented on src/Files.App/Data/Commands/HotKey/HotKey.cs:
```diff
@@ -332,9 +332,21 @@ private static string GetLocalizedKey(string key)
+			var virtualKey = (uint)key;
+			var scanCode = Win32PInvoke.MapVirtualKey(virtualKey, 0);
+
+			if (Win32PInvoke.ToUnicode(virtualKey, scanCode, state, buffer, buffer.Capacity, 0) > 0)
+			{
+				return buffer[^1].ToString();
+			}
```
```suggestion
			if (Win32PInvoke.ToUnicode(virtualKey, scanCode, state, buffer, buffer.Capacity, 0) > 0)
				return buffer[^1].ToString();
```

0x5bfa reviewed (approved):
You know our codebase well!!!
LGTM. code wise.
As we are not in an area that needs unicode support, could you please send a screenshot or video recording of the validation?

XTorLukas commented:
@yaira2 I found the cause of the NumKeys mapping not working, Not localized Pad0-9
![image](https://github.com/files-community/Files/assets/84145589/812f2711-e0d0-4312-a81f-96ce2a78364b)
![image](https://github.com/files-community/Files/assets/84145589/29f5be5b-036a-409c-9879-ed25568e3312)

XTorLukas commented:
### Oh, Decimal (NumPad) key not defined too
![image](https://github.com/files-community/Files/assets/84145589/08a0ec0b-4932-4f89-b3a3-d71932eeb2c6)

XTorLukas commented:
### Test own layout (cs-CZ_QWERTZ)
![Qwertz_cz svg (2)](https://github.com/files-community/Files/assets/84145589/7d4dea9b-ecc2-48f1-af16-5af10b7f6c06)

* **All** Green keys are correct
* **Yellow** keys unused
* **Red keys** (Caps Lock / Scroll Lock / Num Lock) would be good to disable in the selection, because they are not possible, but they are marked as Empty string and so can be overlooked. After saving the application fails.

XTorLukas commented:
> ### Test own layout (cs-CZ_QWERTZ)
> ![Qwertz_cz svg (2)](https://private-user-images.githubusercontent.com/84145589/331682243-7d4dea9b-ecc2-48f1-af16-5af10b7f6c06.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTU5ODMzMTgsIm5iZiI6MTcxNTk4MzAxOCwicGF0aCI6Ii84NDE0NTU4OS8zMzE2ODIyNDMtN2Q0ZGVhOWItZWNjMi00OGYxLWFmMTYtNWFmMTBiN2Y2YzA2LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA1MTclMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNTE3VDIxNTY1OFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWZiYmYxYjBmMGJlZjAyMThiMjY5NGRhZjk2MzA0ZmNkNGM2M2I4ZTViNzA3OGNkNWQxZTNmODRmMWExYWI4OTEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.X-a_oxnkpeXxOGwE_JwOkY6zipF36kMBc6rFMjoaniA)
> 
> * **All** Green keys are correct
> * **Yellow** keys unused
> * **Red keys** (Caps Lock / Scroll Lock / Num Lock) would be good to disable in the selection, because they are not possible, but they are marked as Empty string and so can be overlooked. After saving the application fails.

### Now ignored non-valid keys aka **Red keys** and more optimalization for speed

XTorLukas commented:
### Let's get to testing
- [ ] Testing functionality across different locations
  - [ ] cs-CZ_QWERTZ Layout
  - [ ] en-US_QWERTY Layout
  - [ ]  ... more


- [ ] Verification and search for possible bugs
- [ ] Finalization and verification

XTorLukas commented:
I added information about invalid key binding to give the user the option to use a different key binding.
![image](https://github.com/files-community/Files/assets/84145589/a0763c96-84e9-42d4-9f47-75292d881672)

XTorLukas commented:
### Let's get to testing
- [ ] Testing functionality across different locations
  - [x] `cs-CZ`
  - [x] `sk-SK`
  - [ ] `en-US`
  - [ ]  ... more


- [x] Verification and search for possible bugs
- [ ] Finalization

yaira2 commented on src/Files.App/Views/Settings/ActionsPage.xaml:
```diff
@@ -410,5 +411,13 @@
 			Subtitle="{helpers:ResourceString Name=KeybindingAlreadyUsedNotification}" />
 
+		<!--  Invalid Key TeachingTip  -->
+		<TeachingTip
```
Can you send a screenshot? Also, is there a way I can display this?

XTorLukas replied:
https://github.com/files-community/Files/pull/15420#issuecomment-2118429334

XTorLukas replied:
If pressed down one of this or if not localised aka it's string.Empty
```csharp
// Define invalid keys that shouldn't be processed
var invalidKeys = new HashSet<VirtualKey>
{
	VirtualKey.CapitalLock,
	VirtualKey.NumberKeyLock,
	VirtualKey.Scroll,
};
```

https://github.com/files-community/Files/assets/84145589/5b6f9aba-7829-4ce7-8673-29b1045eb05b

yaira2 replied:
Thanks

yaira2 commented on src/Files.App/Strings/en-US/Resources.resw:
```diff
@@ -3736,6 +3736,9 @@
     <value>This key binding is already being used, please choose a different key binding to continue.</value>
   </data>
+  <data name="KeybindingInvalidKeyNotification" xml:space="preserve">
+    <value>The key bind you pressed cannot be used, try using a different key bind.</value>
```
```suggestion
    <value>This key binding cannot be used, please try again using a different key binding.</value>
```

0x5bfa reviewed (approved):
LGTM other than two things below, code wise.
I can test these on my end this weekend.

0x5bfa reviewed (approved):
Code Quality is over A+.
Looks great to me.

I haven't tested on my end but if someone did I think this can be merged.
If still opened, I'll test this weekend.

XTorLukas commented:
> Code Quality is over A+. Looks great to me.

Thank you, I'm glad I'm not a useless liability here even though I've only been in this community for a short time. I'll definitely keep trying to improve the quality of the code.

XTorLukas commented:
I just have an idea for the future, it would be good to merge all the work with "Key binding" into one class and split it into functions, at least now it's too much for one method 
https://github.com/files-community/Files/blob/d015e048c50baf19fa95f537aa3913c49bc4beee/src/Files.App/Views/Settings/ActionsPage.xaml.cs#L54-L143

0x5bfa commented:
Definitely.
We might want to use MS.Windows.Interactivity and MS.Window.Interaction.Core or just call ViewModel.HandleEditorInput();

XTorLukas commented:
Sounds good.

yaira2 commented:
Will this have any effect on existing users?

XTorLukas commented:
> Will this have any effect on existing users?

It won't be, I just fixed the correct display of mapped keys and added an informative teaching tip about invalid key binding
